cmd/callhome.go

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2015-2022 MinIO, Inc.
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/minio/madmin-go"
+	"github.com/minio/minio/internal/logger"
+)
+
+const (
+	// CallhomeSchemaVersion1 is callhome schema version 1
+	CallhomeSchemaVersion1 = "1"
+
+	// CallhomeSchemaVersion is current callhome schema version.
+	CallhomeSchemaVersion = CallhomeSchemaVersion1
+
+	// CallhomeCycleDefault is the default interval between two callhome cycles (24hrs)
+	CallhomeCycleDefault = 24 * time.Hour
+)
+
+// CallhomeInfo - Contains callhome information
+type CallhomeInfo struct {
+	SchemaVersion string             `json:"schema_version"`
+	AdminInfo     madmin.InfoMessage `json:"admin_info"`
+}
+
+var (
+	enableCallhome            = &safeBool{b: false}
+	callhomeLeaderLockTimeout = newDynamicTimeout(30*time.Second, 10*time.Second)
+	callhomeFreq              = &safeDuration{t: CallhomeCycleDefault}
+)
+
+type safeBool struct {
+	sync.Mutex
+	b bool
+}
+
+func (s *safeBool) Update(b bool) {
+	s.Lock()
+	defer s.Unlock()
+	s.b = b
+}
+
+func (s *safeBool) Get() bool {
+	s.Lock()
+	defer s.Unlock()
+	return s.b
+}
```

Comment:
We might as well make it lockless:

```suggestion
// safeBool contains an atomic bool value.
// Zero value is "false", use safeBoolTrue to initialize a true value.
type safeBool int32

const (
	// Use for initializing only, not for comparisons
	safeBoolFalse safeBool = 0
	safeBoolTrue  safeBool = 1
)

const (
	// Use for initializing only, not for comparisons
	safeBoolFalse safeBool = 0
	safeBoolTrue  safeBool = 1
)

// Update will update the bool.
// Order of concurrent updates is not guaranteed
func (s *safeBool) Update(b bool) {
	if b {
		atomic.StoreInt32((*int32)(s), 1)
	} else {
		atomic.StoreInt32((*int32)(s), 0)
	}
}

// Get returns the bool value.
func (s *safeBool) Get() bool {
	return atomic.LoadInt32((*int32)(s)) == 1
}
```

You are welcome to do the same to `safeDuration` (hint: time.Duration is an int64)

Reply:
Thanks. Have used this, and tried to do the same for `safeDuration` as well. PTAL.